RO-2070: Fått opp farten på søk etter observasjoner

gruble opened the pull request:
Har gjort flg. for å få opp ytelsen på søk etter observasjoner i kartet:

1. Det viste seg at appen fortsatt prøvde å hente observasjoner til offline-bruk selv på web. I tillegg til at-a-glance-søk ble ObservationService.updateObservations() kjørt hver gang man utvidet tidsrommet for observasjoner. Denne trigget søk etter fullstendige observasjoner og kræsjet indexedDB når det returnerte, antakelig pga. for mye data.
Det er mulig hele ObservationService bør fjernes, men jeg lar den ligge inntil vi er ferdig med offline-lagring av observasjoner.
3. Fortsatt ble søk kjørt 3 ganger ved oppstart av app. Ble kvitt et ekstra søk ved å fjerne unødvendig oppdatering av MapService.mapView$ pga. første tikk på FullscreenService.fullScreen$. Fikk også et "tikk" på searchCriteria$ før vi hadde fått et kartutsnitt. Valgte å vente til kartutsnittet var klart før jeg leste searchCriteria. **Litt usikker på løsning der.**

Har også gjort et lite stunt på oppfrisk-knapp og visning av sist-oppdatert-tidspunkt i filter-menyen.

Jeg har en del logging på tidsforbruk og håndering av filter. Dette kan vi sikkert ta vekk etterhvert når vi har spikra løsning.

- [x] Skal vi filtrere på kartutsnitt eller ikke? Det tar 6.5 sek å søke etter siste års observasjoner for Norge fastland. Hvis vi ikke filtrerer på kartutsnitt, tar det 3-4 sek. Fordelen ved å ikke bruke kartutsnitt er at vi slipper å søke hver gang dette endres. Ulempen er at vi ikke får med nye eller endrede observasjoner uten å skrive ekstra kode for å håndtere dette. Jeg valgte å beholde søk på kartusnitt, men lar være å søke på nytt når vi zoomer inn.
- [x] Vise bruker at vi henter data, siden dette tar en del tid. Jeg venter med dette, siden det ikke tar såå lang tid lengre. Men sikkert fint å ha. Har laget en egen oppgave på dette, som vi kan ta seinere: https://nveprojects.atlassian.net/browse/RO-2119 
- [x] Skrive om tegning av observasjoner i kartet slik at det blir like raskt som regobs.no. Jeg fikk tips av Jørgen om å sammenligne tegning av observasjoner i kartet mot regobs.no, og her var det stor forskjell i ytelse. Regobs.no bruker maks 1/2 sekund på å tegne over 13.000 observasjoner, mens beta.regobs.no bruker over 10 sekunder. Dette er løst i egen PR: https://github.com/NVE/regObs4/pull/400

gruble committed:
Fjernet kall som trigget fullt søk etter obser .

I tillegg til at-a-glance-søk ble ObservationService.updateObservations() kjørt hver gang man utvidet tidsrommet for observasjoner. Denne trigget fullt søk og kræsjet indexedDB når det returnerte, antakelig pga. mye data

gruble committed:
Redusert antall søk på forsida

gruble committed:
Kjørt prettier:fix

jorgkv commented:
Setter denne til draft siden den har endringer i en annen PR som evt skal inn først osv.

gruble committed:
RO-2070: Tegner observasjoner i kart slik vi gjør på regobs.no (#400)

* Tegner observasjoner i kart slik vi gjør på regobs.no, så kopiert en del kode fra regobs.no
* Bruker nå CSS for å framheve valgt observasjon. Støtter også "hover" på web

gruble committed:
Fjernet kode som ikke brukes lengre

gruble committed:
Ikke hente observasjoner på nytt ved zoom inn

jorgkv commented on src/app/core/services/search-criteria/search-criteria.service.ts:
```diff
@@ -38,6 +39,16 @@ const URL_PARAM_NICKNAME = 'nick';
 const URL_PARAM_ORDER_BY = 'orderBy';
 const URL_PARAM_ARRAY_DELIMITER = '~'; //https://www.rfc-editor.org/rfc/rfc3986#section-2.3
 
+export function withinExtentCriteriaToBounds(extent: WithinExtentCriteriaDto): L.LatLngBounds {
```
Hvorfor ligger disse funksjonene her, hvis de ikke brukes i servicen?

gruble replied:
Nja, de kunne vel likegodt ligget i en egen util-sak. Det var vel fordi denne servicen har ansvar for å håndtere søkekriteria, så jeg syntes det var logisk. Men sikkert ryddigere å ha den et annet sted, skal fikse

gruble replied:
Har flytta dem til HomePage nå siden det er kun denne sida som bruker dem.

jorgkv commented on src/app/core/services/search-registration/search-registration.service.ts:
```diff
@@ -160,6 +175,13 @@ export class PagedSearchResult<TViewModel> {
 })
 export class SearchRegistrationService {
+  private searchRequested = new Subject<void>();
+  private lastFetched = new Subject<Date>();
```
Jeg tror ikke denne servicen bør holde styr på når det sist ble søkt osv. lastFetched gir bare mening sammen med hvilke søkekriterer som ble brukt. lastFetched for kartvisningen vil alltid være noe annet enn for listevisningen og mine observasjoner, siden andre søkekriterier / andre søkemetoder har blitt brukt.

Vi kan også risikere at man ser på observasjoner i kartet og lastFetched er oppdatert, men så går man til listevisninga, men da har man mista nett og får ikke søkt. Da vil det stå en sist oppdatert dato selv om man ikke får sett på observasjoner, som vil virke rart for brukerne tror jeg? 

Tror derfor hver enkelt side der det søkes, kartsiden, listevisningen, mine observasjoner (hvis det er aktuelt der også), selv må holde styr på dette. Akkurat hvordan vi skal løse det vet jeg ikke helt, men det kunne kanskje vært noe a la dette:

```typescript
class LastFetchedService {

  setLastFetched(...) {
    // Denne metoden kan kartvisningen, listevisningen
    // evt andre steder oppdatere med sin sist oppdatert dato
    // etter man har byttet visning
  }

  setUpdateSearchFunc(..) {
    // Denne metoden kan kartvisningen, listevisningen osv kalle for
    // å gi beskjed om hvilken metode som skal kalles når man trykker
    // på offrisk-knappen
  }

}
```

Så bruker oppfrisk-knapp-komponenten denne servicen.

Et alternativ tror jeg er å ikke vise sist oppdatert dato på web, kun hvis det er brukt offline-søk.

gruble replied:
Bra fanget opp! Jeg synes løsningsforslaget ditt var bra. Prøver å gjøre det på den måten.

gruble replied:
Har laget en egen service for dette som du foreslo. Dette funker sånn noenlunde. Jeg har problemer med to ting:

1. Oppfrisk-knapp i filterpanel når kartfanen er aktiv, trigger ikke søk på nytt, selv om jeg ser at HomePage får det med seg. Ser dette i loggen når jeg trykker oppfrisk-knappen når listefanen er aktiv.
2. Spinner var ikke helt til å stole på, så jeg har tatt den bort midlertidig og nøyer meg med å fjerne "oppdatert-tidspunkt" når man trykker på oppfrisk-knappen.

jorgkv commented on src/app/modules/map/components/map/map.component.ts:
```diff
@@ -348,7 +348,7 @@ export class MapComponent implements OnInit, OnDestroy, AfterViewInit {
     });
 
-    this.fullscreenService.isFullscreen$.pipe(takeUntil(this.ngDestroy$)).subscribe(() => {
+    this.fullscreenService.isFullscreen$.pipe(skip(1), takeUntil(this.ngDestroy$)).subscribe(() => {
```
Hvorfor `skip(1)` her? Gjerne legg til en kommentar, eller hvis `isFullscreen$` ikke fungerer som tenkt, bør det heller løses der?

gruble replied:
Godt poeng. Vi har vel ikke bruk for at isFullscreen$ begynner å mase før man faktisk velger fullskjerm, så jeg løser det der

gruble replied:
Har gjort flere forsøk i FullscreenService på å vente med å sende ticks på isFullscreen$. Det funker forsåvidt, men av en eller annen grunn fører dette til at det blir trigget flere søk etter observasjoner under oppstart. Skulle gjerne visst hvorfor.
Løsningen nå er å la være å skippe første tick på isFullscreen$ i MapComponent. Dette gjør at vi får flere ticks på MapService.mapView$, men siden vi i HomePage uansett venter til vi får searchCriteria$ med et skikkelig kartutsnitt, trigges ikke flere søk under oppstart. Jeg løser det på denne måten i linje 176 i HomePage:
`await firstValueFrom(this.mapService.relevantMapChangeWithInitialView$)`
Hadde jo vært gunstig å kunne fyre av søket tidligere, men dette er min løsning inntil vi finner på noe bedre.

jorgkv reviewed (changes requested):
Har sett gjennom deler av endringene, ser på resten når du har sett på kommentarene.

gruble committed:
Flytta oppfriskningsfunksjon til egen tjeneste

gruble committed:
Flytta konverteringsfunksjoner og fjernet skip(1)

widforss commented on src/app/pages/home/home.page.ts:
```diff
@@ -117,41 +131,89 @@ export class HomePage extends RouterPage implements OnInit, AfterViewChecked {
+          const currentBounds = withinExtentCriteriaToBounds(current.Extent);
+          if (previousBounds.contains(currentBounds)) {
+            this.loggingService.debug('Extent inside previous extent, no need to fetch observations again', DEBUG_TAG);
+            return false; //will stop this criteria change to propagate when we zoom in
```
Her går det å forbedre (hvis man vil). Hvis jeg navigerer sånn her (`[xmin, ymin, xmax, ymax]`):
1. `[0, 0, 100, 100]`
2. `[50, 50, 90, 90]`
3. `[40, 40, 80, 80]`
4. `[80, 40, 120, 80]`

Så vil det i nåværende løsning skje et søk allerede på steg 3, selv om det ikke er nødvendig før vi går ut fra vår originale bbox som vi hadde i steg 1.

gruble replied:
Bra forslag! Jeg har fikset det nå.

widforss commented on src/app/modules/side-menu/components/update-observations/update-observations.component.html:
```diff
@@ -1,9 +1,10 @@
-<ion-item (click)="updateOrCancelObservations()">
+<ion-item (click)="refresh()">
```
Som jeg skjønner koden her er det eneste som skjer når jeg klikker på knappen her er at du kjører .next() på en tom Observable? Det hentes i hvert fall ikke noenting.

widforss reviewed (changes requested):
Ser bra ut. To ting, refresh()-greia føles alvorligere enn zoom-søket, som er mer frivillig.

gruble committed:
Fikset oppfrisk-knappen på kartsida

den virket ikke før vi byttet fane fordi HomePage ikke hadde fått med seg første tick på selectedTab$

gruble committed:
Smartere algoritme for å huske kartutsnitt
